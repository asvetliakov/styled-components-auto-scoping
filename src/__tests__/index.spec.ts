import plugin from "../index";
import { transform } from "@babel/core";

it("Moves styled component to outer scope", () => {
    const src = `
function Test1() {
    const A1 = styled.div\`
        color: #fff;
    \`;

    const A2 = styled(A1)\`
        color: #000;
    \`;

    const A3 = styled("div")\`
        color: #fff;
    \`;
}

const Test2 = () => {
    const A4 = styled.div\`
        color: #fff;
    \`;

    const A5 = styled(A4)\`
        color: #000;
    \`;

    const A6 = styled("div")\`
        color: #fff;
    \`;
}

class Test3 extends Component {
    render() {
        const A7 = styled.div\`\`;
        const A8 = styled(Test2)\`\`;
    }
}
    `;

    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin] })!.code).toMatchSnapshot();
});

it("Does not transform top level styled components", () => {
    const src = `
const A = styled.div\`\`;
const B = styled.div\`
    \${props => props.a};
\`;

const C = styled(A)\`\`;
const D = styled(Comp)\`
    \${someProp};
\`;
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin] })!.code).toMatchSnapshot();
});

it("Process local identifiers in styled templates", () => {
    const src = `
import { themeProp } from "theme";
const someVar = 5;

function Test1(props) {
    const { localProp } = props;
    const A = styled.div\`
        color: green;
        color: \${props};
        color: \${localProp};
        color: \${someVar};
        color: \${themeProp};
        color: \${nonExisted};
        background: \${localProp};
    \`;

    return (
        <div>
            <A prop="test"/>
            <A>test</A>
        </div>
    );
}

class Test2 extends Component {
    render() {
        const { clProp } = this.props;
        const B = styled.div\`
            color: green;
            color: \${clProp};
            color: \${someVar};
            color: \${themeProp};
            color: \${nonExisted};
            background: \${clProp};
        \`;

        return (
            <div>
                <B/>
                <B>test</B>
            </div>
        )
    }
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin], presets: ["@babel/react"] })!.code).toMatchSnapshot();
});

it("Process local member access in styled templates", () => {
    const src = `
import { theme } from "theme";
const someObj = { a: true };

function Test1(props) {
    const { localProp } = props;
    const A = styled.div\`
        color: \${props.a};
        color: \${props.b.c};
        color: \${localProp};
        color: \${localProp.e};
        color: \${theme.a};
        color: \${someObj.a};
        background: \${props.a};
        background: \${localProp.e};
    \`;

    return (
        <div>
            <A prop="test"/>
            <A>test</A>
        </div>
    );
}

class Test2 extends Component {
    render() {
        const { clProp } = this.props;
        const B = styled.div\`
            color: \${clProp.a};
            color: \${clProp.a.b};
            color: \${this.props};
            color: \${this.props.c};
            color: \${theme.a};
            color: \${someObj.a};
            background: \${clProp.a};
            background: \${clProp.d};
            background: \${this.props.c};
            background: \${this.props.k};
        \`;

        return (
            <div>
                <B/>
                <B>test</B>
            </div>
        )
    }
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin], presets: ["@babel/react"] })!.code).toMatchSnapshot();

});

it("Process logical expressions in styled templates", () => {
    const src = `
import { theme } from "theme";

function Test1(props) {
    const { localProp } = props;
    const A = styled.div\`
        color: \${props.a && "green"};
        color: \${localProp.b && props.a1};
        color: \${localProp.c ? props.x : props.y};
        color: \${localProp.c && localProp.k ? props.e : props.t.t ? props.m.m : "test"};
        color: \${props.a && theme.a};
        color: \${props.a ? theme.a : theme.b};
    \`;

    return (
        <A />
    );
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin], presets: ["@babel/react"] })!.code).toMatchSnapshot();
});

it("Process arrow function expressions in styled templates", () => {
    const src = `
import { theme } from "theme";
const someObj = { a: true };

function Test1(props) {
    const { localProp, anotherProp, xyz } = props;
    const A = styled.div\`
        color: \${props => props.a && localProp};
        color: \${props => props.a && localProp ? anotherProp : "Test" };
        color: \${differentName => differentName.a && xyz};
    \`;

    return (
        <A a={props.a} />
    );
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin], presets: ["@babel/react"] })!.code).toMatchSnapshot();
});

it("Adds attribute name prefix when specified", () => {
    const src = `

function Test1(props) {
    const { localProp } = props;
    const A = styled.div\`
        color: \${localProp};
        color: \${props.b};
    \`;

    return (
        <A prop="test"/>
    );
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [[plugin, { addAttributePrefix: "ui_" }]], presets: ["@babel/react"] })!.code).toMatchSnapshot();
});

it("Works for negative logical expressions", () => {
    const src = `
import { theme } from "theme";

function Test1(props) {
    const { localProp } = props;
    const A = styled.div\`
        color: \${!props.a && "green"};
        color: \${!localProp.b && props.a1};
        color: \${!localProp.c ? props.x : props.y};
        color: \${!localProp.c && localProp.k ? props.e : props.t.t ? props.m.m : "test"};
        color: \${!props.a && theme.a};
        color: \${!props.a ? theme.a : theme.b};
    \`;

    return (
        <A />
    );
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin], presets: ["@babel/react"] })!.code).toMatchSnapshot();
});

it("Works with styled.attrs", () => {
    const src = `
function Test1() {
    const A1 = styled.div.attrs({ className: "test" })\`
        color: #fff;
    \`;
    const A2 = styled(A1).attrs({ className: "test2" })\`
        color: #fff;
    \`;
}
    `;
    expect(transform(src, { babelrc: false, configFile: false, plugins: [plugin] })!.code).toMatchSnapshot();
});