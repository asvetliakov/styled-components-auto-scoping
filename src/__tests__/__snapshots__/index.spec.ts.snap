// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adds attribute name prefix when specified 1`] = `
"const A = styled.div\`
        color: \${p => p.ui_localProp};
        color: \${p => p.ui_b};
    \`;

function Test1(props) {
  const {
    localProp
  } = props;
  return React.createElement(A, {
    prop: \\"test\\",
    ui_localProp: localProp,
    ui_b: props.b
  });
}"
`;

exports[`Does not transform top level styled components 1`] = `
"const A = styled.div\`\`;
const B = styled.div\`
    \${props => props.a};
\`;
const C = styled(A)\`\`;
const D = styled(Comp)\`
    \${someProp};
\`;"
`;

exports[`Moves styled component to outer scope 1`] = `
"const A1 = styled.div\`
        color: #fff;
    \`;
const A2 = styled(A1)\`
        color: #000;
    \`;
const A3 = styled(\\"div\\")\`
        color: #fff;
    \`;

function Test1() {}

const A4 = styled.div\`
        color: #fff;
    \`;
const A5 = styled(A4)\`
        color: #000;
    \`;
const A6 = styled(\\"div\\")\`
        color: #fff;
    \`;

const Test2 = () => {};

const A7 = styled.div\`\`;
const A8 = styled(Test2)\`\`;

class Test3 extends Component {
  render() {}

}"
`;

exports[`Process arrow function expressions in styled templates 1`] = `
"import { theme } from \\"theme\\";
const someObj = {
  a: true
};
const A = styled.div\`
        color: \${props => props.a && props.localProp};
        color: \${props => props.a && props.localProp ? props.anotherProp : \\"Test\\"};
        color: \${differentName => differentName.a && differentName.xyz};
    \`;

function Test1(props) {
  const {
    localProp,
    anotherProp,
    xyz
  } = props;
  return React.createElement(A, {
    a: props.a,
    localProp: localProp,
    anotherProp: anotherProp,
    xyz: xyz
  });
}"
`;

exports[`Process local identifiers in styled templates 1`] = `
"import { themeProp } from \\"theme\\";
const someVar = 5;
const A = styled.div\`
        color: green;
        color: \${p => p.props};
        color: \${p => p.localProp};
        color: \${someVar};
        color: \${themeProp};
        color: \${nonExisted};
        background: \${p => p.localProp};
    \`;

function Test1(props) {
  const {
    localProp
  } = props;
  return React.createElement(\\"div\\", null, React.createElement(A, {
    prop: \\"test\\",
    props: props,
    localProp: localProp
  }), React.createElement(A, {
    props: props,
    localProp: localProp
  }, \\"test\\"));
}

const B = styled.div\`
            color: green;
            color: \${p => p.clProp};
            color: \${someVar};
            color: \${themeProp};
            color: \${nonExisted};
            background: \${p => p.clProp};
        \`;

class Test2 extends Component {
  render() {
    const {
      clProp
    } = this.props;
    return React.createElement(\\"div\\", null, React.createElement(B, {
      clProp: clProp
    }), React.createElement(B, {
      clProp: clProp
    }, \\"test\\"));
  }

}"
`;

exports[`Process local member access in styled templates 1`] = `
"import { theme } from \\"theme\\";
const someObj = {
  a: true
};
const A = styled.div\`
        color: \${p => p.a};
        color: \${p => p.c};
        color: \${p => p.localProp};
        color: \${p => p.e};
        color: \${theme.a};
        color: \${someObj.a};
        background: \${p => p.a};
        background: \${p => p.e};
    \`;

function Test1(props) {
  const {
    localProp
  } = props;
  return React.createElement(\\"div\\", null, React.createElement(A, {
    prop: \\"test\\",
    a: props.a,
    c: props.b.c,
    localProp: localProp,
    e: localProp.e
  }), React.createElement(A, {
    a: props.a,
    c: props.b.c,
    localProp: localProp,
    e: localProp.e
  }, \\"test\\"));
}

const B = styled.div\`
            color: \${p => p.a};
            color: \${p => p.b};
            color: \${p => p.props};
            color: \${p => p.c};
            color: \${theme.a};
            color: \${someObj.a};
            background: \${p => p.a};
            background: \${p => p.d};
            background: \${p => p.c};
            background: \${p => p.k};
        \`;

class Test2 extends Component {
  render() {
    const {
      clProp
    } = this.props;
    return React.createElement(\\"div\\", null, React.createElement(B, {
      a: clProp.a,
      b: clProp.a.b,
      props: this.props,
      c: this.props.c,
      d: clProp.d,
      k: this.props.k
    }), React.createElement(B, {
      a: clProp.a,
      b: clProp.a.b,
      props: this.props,
      c: this.props.c,
      d: clProp.d,
      k: this.props.k
    }, \\"test\\"));
  }

}"
`;

exports[`Process logical expressions in styled templates 1`] = `
"import { theme } from \\"theme\\";
const A = styled.div\`
        color: \${p => p.a && \\"green\\"};
        color: \${p => p.b && p.a1};
        color: \${p => p.c ? p.x : p.y};
        color: \${p => p.c && p.k ? p.e : p.t ? p.m : \\"test\\"};
        color: \${p => p.a && theme.a};
        color: \${p => p.a ? theme.a : theme.b};
    \`;

function Test1(props) {
  const {
    localProp
  } = props;
  return React.createElement(A, {
    a: props.a,
    b: localProp.b,
    a1: props.a1,
    c: localProp.c,
    x: props.x,
    y: props.y,
    k: localProp.k,
    e: props.e,
    t: props.t.t,
    m: props.m.m
  });
}"
`;

exports[`Works for negative logical expressions 1`] = `
"import { theme } from \\"theme\\";
const A = styled.div\`
        color: \${p => !p.a && \\"green\\"};
        color: \${p => !p.b && p.a1};
        color: \${p => !p.c ? p.x : p.y};
        color: \${p => !p.c && p.k ? p.e : p.t ? p.m : \\"test\\"};
        color: \${p => !p.a && theme.a};
        color: \${p => !p.a ? theme.a : theme.b};
    \`;

function Test1(props) {
  const {
    localProp
  } = props;
  return React.createElement(A, {
    a: props.a,
    b: localProp.b,
    a1: props.a1,
    c: localProp.c,
    x: props.x,
    y: props.y,
    k: localProp.k,
    e: props.e,
    t: props.t.t,
    m: props.m.m
  });
}"
`;

exports[`Works with styled.attrs 1`] = `
"const A1 = styled.div.attrs({
  className: \\"test\\"
})\`
        color: #fff;
    \`;
const A2 = styled(A1).attrs({
  className: \\"test2\\"
})\`
        color: #fff;
    \`;

function Test1() {}"
`;
